scipy/cluster/hierarchy.py

```diff
@@ -2251,10 +2251,12 @@ def is_valid_linkage(Z, warning=False, throw=False, name=None):
             if xp.any(Z[:, 3] > (Z.shape[0] + 1)):
                 raise ValueError('Linkage matrix contains excessive observations'
                                  'in a cluster')
-        if _check_hierarchy_uses_cluster_before_formed(Z):
+        if xp.any(
+            xp.max(Z[:, :2], axis=1) >= xp.arange(n + 1, 2 * n + 1, dtype=Z.dtype)
+        ):
             raise ValueError(f'Linkage {name_str}uses non-singleton cluster before'
                              ' it is formed.')
-        if _check_hierarchy_uses_cluster_more_than_once(Z):
+        if xp.unique_values(Z[:, :2]).shape[0] < n * 2:
```

Comment:
TODO change to `xpx.nunique` for Dask support (requires array-api-extra 0.7.0).

Reply:
That can be done in another PR right, no change needed here?